File: tests/python/extensions/pybind/core/attitude.py
import copy
import h5py
import numpy as np
import numpy.testing as npt
from isce3.ext.isce3 import core
from tempfile import mkstemp

def test_attitude():
    n = 11
    t = np.linspace(0, 1, n)
    epoch = core.DateTime(2023, 1, 1)

    # All quaternions no rotation.
    q = [core.Quaternion(1, 0, 0, 0) for i in range(n)]
    attitude = core.Attitude(t, q, epoch)

    def rotmat(t):
        return attitude.interpolate(t).to_rotation_matrix()

    ti = 0.5
    assert np.allclose(rotmat(ti), np.eye(3))

    # Make sure we don't crash at endpoints.
    assert np.allclose(rotmat(t[0]), np.eye(3))
    assert np.allclose(rotmat(t[-1]), np.eye(3))

    # Varying rotation around constant axis.  No rotation at time ti.
    np.random.seed(1234)
    axis = np.random.randn(3)
    axis /= np.linalg.norm(axis)
    angles = np.linspace(-1, 1, n)
    w = np.cos(angles / 2)
    xyz = np.sin(angles / 2)[:,None] * axis[None,:]
    q = [core.Quaternion(w[i], *xyz[i]) for i in range(n)]

    attitude = core.Attitude(t, q, epoch)
    assert np.allclose(rotmat(ti), np.eye(3))

    # Check that rotation axis remains invariant with interpolation.
    ni = 1 + 32 * (n - 1)
    for x in np.linspace(t[0], t[-1], ni):
        R = rotmat(x)
        assert np.allclose(R.dot(axis), axis)


def dummy_attitude():
    t = [0.0, 0.1]
    q = (core.Quaternion(1,0,0,0),) * 2
    epoch = core.DateTime(2020, 1, 1)
    return core.Attitude(t, q, epoch)


def test_properties_exist():
    attitude = dummy_attitude()
    attitude.size
    attitude.time
    attitude.quaternions
    attitude.start_time
    attitude.end_time
    attitude.start_datetime
    attitude.end_datetime
    attitude.reference_epoch


def test_update_epoch():
    attitude = dummy_attitude()
    i = -1
    old_epoch = attitude.reference_epoch
    old_timestamp = old_epoch + core.TimeDelta(attitude.time[i])

    new_epoch = old_epoch + core.TimeDelta(100.0)
    attitude.update_reference_epoch(new_epoch)
    assert attitude.reference_epoch == new_epoch

    new_timestamp = attitude.reference_epoch + core.TimeDelta(attitude.time[i])
    assert (new_timestamp - old_timestamp).total_seconds() < 1e-9


def test_io():
    attitude = dummy_attitude()
    _, name = mkstemp()
    # write
    with h5py.File(name, "w") as h5:
        g = h5.create_group("/attitude")
        attitude.save_to_h5(g)
    # read
    with h5py.File(name, "r") as h5:
        core.Attitude.load_from_h5(h5["/attitude"])


def test_copy():
    attitude = dummy_attitude()
    # only modifiable attribute via python is epoch
    epoch = attitude.reference_epoch + core.TimeDelta(1.0)
    for a in (copy.copy(attitude), copy.deepcopy(attitude), attitude.copy()):
        a.update_reference_epoch(epoch)
        assert a.reference_epoch != attitude.reference_epoch


def test_contains():
    attitude = dummy_attitude()
    assert not attitude.contains(attitude.start_time - 1.0)
    assert not attitude.contains(attitude.end_time + 1.0)
    mid = 0.5 * (attitude.start_time + attitude.end_time)
    assert attitude.contains(mid)


def test_crop():
    t = [0, 1, 2, 3, 4, 5]
    q = (core.Quaternion(1,0,0,0),) * len(t)
    epoch = core.DateTime(2020, 1, 1)
    attitude = core.Attitude(t, q, epoch)

    # Pick a point past 2nd sample and before 5th sample.  Expect to only lose
    # the samples at the end so that the requested points are contained in the
    # output domain.
    t0, t1 = 1.1, 3.9
    start = epoch + core.TimeDelta(t0)
    stop = epoch + core.TimeDelta(t1)
    cropped_attitude = attitude.crop(start, stop)

    npt.assert_equal(cropped_attitude.size, len(t) - 2)
    npt.assert_(cropped_attitude.contains(t0))
    npt.assert_(cropped_attitude.contains(t1))

    # Requesting npad=1 should result in getting all the original data.
    cropped_attitude = attitude.crop(start, stop, npad=1)
    npt.assert_array_equal(attitude.time, cropped_attitude.time)

    # Check that cropping doesn't modify epoch.
    npt.assert_(cropped_attitude.reference_epoch == attitude.reference_epoch)

File: python/extensions/pybind_isce3/core/Attitude.cpp
#include "Quaternion.h"

#include <pybind11/eigen.h>
#include <pybind11/numpy.h>
#include <pybind11/stl.h>

#include <isce3/core/DateTime.h>
#include <isce3/io/IH5.h>
#include <isce3/core/Quaternion.h>
#include <isce3/core/Serialization.h>

namespace py = pybind11;
using namespace isce3::core;

void addbinding(pybind11::class_<Attitude>& pyAttitude)
{
    pyAttitude
        .def(py::init<std::vector<double>, std::vector<Quaternion>, DateTime>(),
            py::arg("time"), py::arg("quaternions"), py::arg("epoch"))
        .def("interpolate", &Attitude::interpolate, py::arg("time"))
        .def_static("load_from_h5", [](py::object h5py_group) {
            auto id = h5py_group.attr("id").attr("id").cast<hid_t>();
            isce3::io::IGroup group(id);
            Attitude att;
            loadFromH5(group, att);
            return att;
        },
        "De-serialize Attitude from h5py.Group object",
        py::arg("h5py_group"))
        .def("save_to_h5", [](const Attitude& self, py::object h5py_group) {
                auto id = h5py_group.attr("id").attr("id").cast<hid_t>();
                isce3::io::IGroup group(id);
                isce3::core::saveToH5(group, self);
            },
            "Serialize Attitude to h5py.Group object.",
            py::arg("h5py_group"))

        // add a function rather than just a settable property since
        // self.time is also modified
        .def("update_reference_epoch",
            py::overload_cast<const DateTime&>(&Attitude::referenceEpoch),
            R"(
            Set reference epoch and modify time tags so that
                self.reference_epoch + TimeDelta(self.time[i])
            remains invariant for all i in range(len(self.time)).)",
            py::arg("reference_epoch"))

        .def("copy", [](const Attitude& self) { return Attitude(self);})
        .def("__copy__", [](const Attitude& self) { return Attitude(self);})
        .def("__deepcopy__", [](const Attitude& self, py::dict) {
                return Attitude(self);
            }, py::arg("memo"))

        .def_property_readonly("size", &Attitude::size)
        .def_property_readonly("time", &Attitude::time)
        .def_property_readonly("quaternions", &Attitude::quaternions)
        .def_property_readonly("start_time", &Attitude::startTime)
        .def_property_readonly("end_time", &Attitude::endTime)
        .def_property_readonly("start_datetime", &Attitude::startDateTime)
        .def_property_readonly("end_datetime", &Attitude::endDateTime)
        .def_property_readonly("reference_epoch",
                py::overload_cast<>(&Attitude::referenceEpoch, py::const_))
        .def("contains", &Attitude::contains,
            "Check if time falls in the valid interpolation domain.",
            py::arg("time"))

        .def("crop", &Attitude::crop,
            R"(
            Create a new Attitude containing data in the requested interval

            Parameters
            ----------
            start : isce3.core.DateTime
                Beginning of time interval
            end : isce3.core.DateTime
                End of time interval
            npad : int, optional
                Minimal number of state vectors to include past each of
                the given time bounds (useful to guarantee adequate
                support for interpolation).
            Returns
            -------
            isce3.core.Attitude
                Attitude object with data containing start & end times)",
            py::arg("start"), py::arg("end"), py::arg("npad") = 0)
        ;
}
